packages/puppeteer-core/src/util/disposable.ts

```
@@ -145,15 +143,40 @@ export class DisposableStack {
    */
   move(): DisposableStack {
     if (this.#disposed) {
-      throw new ReferenceError('a disposed stack can not use anything new'); // step 3
+      throw new ReferenceError('A disposed stack can not use anything new');
     }
-    const stack = new DisposableStack(); // step 4-5
+    const stack = new DisposableStack();
     stack.#stack = this.#stack;
     this.#disposed = true;
     return stack;
   }
 
-  [disposeSymbol] = this.dispose;
+  /**
+   * Disposes each resource in the stack in last-in-first-out (LIFO) manner
+   * @returns {void}
+   */
+  [disposeSymbol](): void {
+    if (this.#disposed) {
+      return;
+    }
+    this.#disposed = true;
+    const errors: unknown[] = [];
+    for (const resource of this.#stack) {
+      try {
+        resource[disposeSymbol]();
+      } catch (e) {
+        errors.push(e);
+      }
+    }
+    if (errors.length === 1) {
```

From my understanding this is not suppose to throw an error, it should suppress the errors that are produced by resouce[disposeSymbol].

---

```
/*
using declarations are supposed to be resilient to exceptions; if an error is thrown, it’s rethrown after disposal. On the other hand, the body of your function might execute as expected, but the Symbol.dispose might throw. In that case, that exception is rethrown as well.

But what happens if both the logic before and during disposal throws an error? For those cases, SuppressedError has been introduced as a new subtype of Error. It features a suppressed property that holds the last-thrown error, and an error property for the most-recently thrown error.
*/

class ErrorA extends Error {
    name = "ErrorA";
}
class ErrorB extends Error {
    name = "ErrorB";
}
function throwy(id: string) {
    return {
        [Symbol.dispose]() {
            throw new ErrorA(`Error from ${id}`);
        }
    };
}
function func() {
    using a = throwy("a");
    throw new ErrorB("oops!")
}
try {
    func();
}
catch (e: any) {
    console.log(e.name); // SuppressedError
    console.log(e.message); // An error was suppressed during disposal.
    console.log(e.error.name); // ErrorA
    console.log(e.error.message); // Error from a
    console.log(e.suppressed.name); // ErrorB
    console.log(e.suppressed.message); // oops!
}
```

According to the above paragraph from [TC39 doc](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-2.html#using-declarations-and-explicit-resource-management) this should throw a normal error only and the language implementation of `using` should handle that this should be converted to a SuppressedError if the code logic before calling disposal also throws error, otherwise in normal cases where either only normal code flow throws error or only disposal throws error we don't have SuppressedError in that case

---

Oh just went through this file as well - https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#the-disposablestack-and-asyncdisposablestack-container-objects 

I think it explain it much better and I see what you are saying, let me update it accordingly as well

---

Fixed now following TC39 spec if it is a single error we throw it as it is, but if we encounter multiple errors then they are wrapped in SuppressedError in the order as specified under TC39 i.e. main error as the first encountered error and remaining error stored as chain of suppressed errors

---

Also just for context linking this paragraph from above TC39 proposal based on which I think error should be thrown - 
```
AsyncDisposableStack is the async version of DisposableStack and is a container used to aggregate async disposables, guaranteeing that every disposable resource in the container is disposed when the respective disposal method is called. If any disposable in the container throws an error during dispose, or results in a rejected Promise, it would be thrown at the end (possibly wrapped in a SuppressedError if multiple errors were thrown):
```